core/src/test/java/org/elasticsearch/action/bulk/RetryTests.java

```diff
@@ -0,0 +1,245 @@
+        AssertingListener listener = new AssertingListener();
+
+        BulkRequest bulkRequest = createBulkRequest();
+        Retry.on(EsRejectedExecutionException.class)
+                .policy(linearBackoff)
+                .withAsyncBackoff(bulkClient, bulkRequest, listener);
+
+        listener.awaitCallbacksCalled();
+        listener.assertOnResponseCalled();
+        listener.assertResponseWithoutFailures();
+        listener.assertResponseWithNumberOfItems(bulkRequest.numberOfActions());
+        listener.assertOnFailureNeverCalled();
+    }
+
+    public void testAsyncRetryFailsAfterBacksOff() throws Exception {
+        BackoffPolicy linearBackoff = ConstantBackoff.times(callsToFail - 1);
+        AssertingListener listener = new AssertingListener();
+
+        BulkRequest bulkRequest = createBulkRequest();
+        Retry.on(EsRejectedExecutionException.class)
+                .policy(linearBackoff)
+                .withAsyncBackoff(bulkClient, bulkRequest, listener);
+
+        listener.awaitCallbacksCalled();
+
+        listener.assertOnResponseCalled();
+        listener.assertResponseWithFailures();
+        listener.assertResponseWithNumberOfItems(bulkRequest.numberOfActions());
+        listener.assertOnFailureNeverCalled();
+    }
+
+    private static class AssertingListener implements ActionListener<BulkResponse> {
+        private final CountDownLatch latch;
+        private int countOnResponseCalled = 0;
+        private Throwable lastFailure;
+        private BulkResponse response;
+
+        private AssertingListener() {
+            latch = new CountDownLatch(1);
+        }
+
+        public void awaitCallbacksCalled() throws InterruptedException {
+            latch.await();
+        }
+
+        @Override
+        public void onResponse(BulkResponse bulkItemResponses) {
+            latch.countDown();
+            this.response = bulkItemResponses;
+            countOnResponseCalled++;
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            latch.countDown();
+            this.lastFailure = e;
+        }
+
+        public void assertOnResponseCalled() {
+            assertThat(countOnResponseCalled, equalTo(1));
+        }
+
+        public void assertResponseWithNumberOfItems(int numItems) {
+            assertThat(response.getItems().length, equalTo(numItems));
+        }
+
+        public void assertResponseWithoutFailures() {
+            assertThat(response, notNullValue());
+            assertFalse("Response should not have failures", response.hasFailures());
+        }
+
+        public void assertResponseWithFailures() {
+            assertThat(response, notNullValue());
+            assertTrue("Response should have failures", response.hasFailures());
+        }
+
+        public void assertOnFailureNeverCalled() {
+            assertThat(lastFailure, nullValue());
+        }
+    }
+
+
+    private static final class ConstantBackoff extends BackoffPolicy {
```

Comment:
Maybe pull this guy into BackoffPolicy? I figure it'd be useful for anyone uses BulkProcessor in tests. You'd want to make the DELAY configurable, but that is pretty simple.